src/sage/graphs/graph_decompositions/slice_decomposition.pyx

```diff
@@ -0,0 +1,1079 @@
+            ValueError: parameter G must be an undirected graph
+
+        .. automethod:: __getitem__
+        """
+        if G.is_directed():
+            raise ValueError("parameter G must be an undirected graph")
+
+        if initial_vertex is not None and initial_vertex not in G:
+            raise LookupError(f"vertex ({initial_vertex}) is not a vertex of the graph")
+
+        cdef CGraphBackend Gbackend = <CGraphBackend> G._backend
+        cdef CGraph cg = Gbackend.cg()
+
+        self._graph_class = type(G)
+
+        cdef int initial_v_int
+        if initial_vertex is not None:
+            # we already checked that initial_vertex is in G
+            initial_v_int = Gbackend.get_vertex(initial_vertex)
+        else:
+            initial_v_int = -1
+
+        cdef vector[int] sigma
+        cdef vector[vector[int]] lex_label
+
+        # Compute the slice decomposition using the extended lexBFS algorithm
+        extended_lex_BFS(cg, sigma, NULL, initial_v_int, NULL,
+                         &(self.xslice_len), &lex_label)
+
+        # Translate the results with the actual vertices of the graph
+        self.sigma = tuple(Gbackend.vertex_label(v_int) for v_int in sigma)
+        self.sigma_inv = {v: i  for i, v in enumerate(self.sigma)}
+        self.lex_label = {i: tuple(Gbackend.vertex_label(v_int) for v_int in lli)
+                                    for i, lli in enumerate(lex_label)}
+
+    def __eq__(self, other):
+        """
+        Return whether ``self`` and ``other`` are equal.
+
+        TESTS::
+
+            sage: G = graphs.PetersenGraph()
+            sage: SD = G.slice_decomposition()
+            sage: SD == SD
+            True
+            sage: SD == G.slice_decomposition()
+            True
+
+            sage: P3 = graphs.PathGraph(3)
+            sage: SD1 = P3.slice_decomposition(initial_vertex=0)
+            sage: SD2 = P3.slice_decomposition(initial_vertex=2)
+            sage: SD1 == SD2
+            False
+            sage: SD3 = graphs.CompleteGraph(3).slice_decomposition()
+            sage: SD1 == SD3  # same lexBFS but different slice for 1
+            False
+            sage: SD4 = Graph([(0,1), (0,2)]).slice_decomposition()
+            sage: SD3 == SD4  # same lexBFS and slices but different active edges
+            False
+        """
+        if not isinstance(other, type(self)):
+            return False
+
+        cdef SliceDecomposition sd = <SliceDecomposition>other
+
+        return self.sigma_inv == sd.sigma_inv \
+                and self.lex_label == sd.lex_label \
+                and self.xslice_len == sd.xslice_len
+
+    def __hash__(self):
+        r"""
+        Compute a hash of a ``SliceDecomposition`` object.
+
+        TESTS::
+
+            sage: P3 = graphs.PathGraph(3)
+            sage: SD1 = P3.slice_decomposition(initial_vertex=0)
+            sage: SD2 = P3.slice_decomposition(initial_vertex=2)
+            sage: len({SD1: 1, SD2: 2})  # indirect doctest
```

Comment:
This seems ok, but why not simply `hash(SD1) != hash(SD2)` ?

Reply:
I got the idea by looking at other doctests for other `__hash__` methods.
By thinking a little about it, there is a nonzero probability that there exists an architecture where the hashes could be equal. But as keys of a dict are compare via `__eq__` if they have the same hases, even in this (very very) improbable case the test should pass.

Reply:
You are right. Thank you for the clarification.